src/schema/mod.rs

```diff
@@ -0,0 +1,944 @@
+//! Logic for parsing and interacting with schemas in Avro format.
+use std::{
+    borrow::Cow,
+    cell::RefCell,
+    collections::{HashMap, HashSet},
+    fmt,
+    rc::Rc,
+};
+
+use digest::Digest;
+use failure::Error;
+use serde_json::{self, Map, Value};
+
+use once_cell::sync::OnceCell;
+use string_interner::{DefaultStringInterner, Sym as RawNameSym};
+use wyhash::{self, WyHash};
+
+use crate::types;
+use crate::util::MapHelper;
+
+pub mod builder;
+mod data;
+mod parse;
+mod pcf;
+mod serialize;
+use crate::schema::builder::SchemaBuilder;
+use crate::schema::data::{RecordFieldData, SchemaData};
+use crate::schema::parse::SchemaParser;
+
+// Rexport the error types as part of the public API
+pub use builder::{BuilderError, BuilderErrors};
+pub use parse::ParseSchemaError;
+
+/// Trait to provide hash implementations for HashMaps
+#[derive(Clone, Default)]
+pub(crate) struct WyHashBuilder;
+
+impl std::hash::BuildHasher for WyHashBuilder {
+    type Hasher = WyHash;
+
+    fn build_hasher(&self) -> Self::Hasher {
+        WyHash::default()
+    }
+}
+
+/// Simple type to reduce noise in definitions
+type JsonMap = Map<String, Value>;
+
+/// Representation of an avro schema
+#[derive(Clone)]
+pub struct Schema {
+    namespace_names: DefaultStringInterner,
+    type_names: DefaultStringInterner,
+    aliases: HashMap<NameRef, NameRef, WyHashBuilder>,
+    reverse_aliases: HashMap<NameRef, Vec<NameRef>, WyHashBuilder>,
+    types: HashMap<NameRef, SchemaData, WyHashBuilder>,
+    root: NameRef,
+}
+
+impl Schema {
+    /// Parse the schema from a serde json value
+    pub fn parse(value: &Value) -> Result<Schema, Error> {
+        SchemaParser::parse(value).map_err(|e| e.into())
+    }
+
+    /// Parse the schema from a string
+    pub fn parse_str(raw: &str) -> Result<Schema, Error> {
+        Self::parse_slice(raw.as_bytes())
+    }
+
+    /// Parse the schema from a slice
+    pub fn parse_slice<'a>(raw: &'a [u8]) -> Result<Schema, Error> {
+        let value = serde_json::from_slice(raw)?;
+        Self::parse(&value).map_err(|e| e.into())
+    }
+
+    /// Create a builder for the construction of a schema
+    pub fn builder() -> SchemaBuilder {
+        SchemaBuilder::new()
+    }
+
+    /// Get the root element for the schema, this is the element that defines the current schema
+    pub fn root(&self) -> SchemaType<'_> {
```

Comment:
I think that the name `root` is befitting of the recursive nature of avro schemas, but the [spec](http://avro.apache.org/docs/current/spec.html) doesn't mention the word at all. I see that it is used in the [C++ implementation](http://avro.apache.org/docs/1.7.7/api/cpp/html/classavro_1_1Schema.html), but I was wondering if something like `schema_type` could be more descriptive...

What's your opinion?

Reply:
I have little opinion, I mostly took root as in root of the tree, I dont mind renaming it to something more descriptive

Reply:
My vote would be for `schema_type`, I think. @flavray , do you have any opinion?